plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.slice.Slice;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.ColumnarRow;
+import io.trino.spi.connector.ConnectorPageSource;
+import io.trino.spi.connector.EmptyPageSource;
+import io.trino.spi.connector.UpdatablePageSource;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Collections.emptyIterator;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Concatenates the Pages for all PageSources which were grouped together an Iceberg CombinedScanTask.
+ */
+public class IcebergCombinedPageSource
+        implements UpdatablePageSource
+{
+    private final Map<String, UpdatablePageSource> delegatePageSources;
+    private final Iterator<UpdatablePageSource> pageSourceIterator;
+
+    private ConnectorPageSource currentPageSource;
+
+    public IcebergCombinedPageSource(Map<String, UpdatablePageSource> delegatePageSourcesByPath)
+    {
+        if (delegatePageSourcesByPath.isEmpty()) {
+            this.delegatePageSources = ImmutableMap.of();
+            this.pageSourceIterator = emptyIterator();
+            this.currentPageSource = new EmptyPageSource();
+        }
+        else {
+            this.delegatePageSources = ImmutableMap.copyOf(requireNonNull(delegatePageSourcesByPath, "delegatePageSources is null"));
+            this.pageSourceIterator = delegatePageSourcesByPath.values().iterator();
+            this.currentPageSource = pageSourceIterator.next();
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegatePageSources.values().stream()
```

Comment:
This recomputes the values for every call. Instead, we can compute as we go, since the stats for completed page sources will not change. See `io.trino.plugin.raptor.legacy.util.ConcatPageSource`